[k8s_cloud_beta1] Adding support for ssh using kubectl port-forward to access k8s instance

landscapepainter opened the pull request:
<!-- Describe the changes in this PR -->
This PR allows to set up an ssh session to access k8s instance through a jump pod using `kubectl port-forward` and `socat` as a `Proxycommand`. Default ssh method would be to use `kubectl port-forward` and `socat`, and if the user wants to open up a `NodePort` service to access the jump pod instead, the following can be written in `~/.sky/config.yaml`:

```
kubernetes:
  networking: nodeport
```

Note: Allowing to use `ControlMaster` that is set from `ssh_options_list()` causes `ControlPersist `amount of seconds of stalling whenever `ssh` command is ran while running `sky` commands in some linux distributions(ubuntu 22.04, debian 11 seems to be okay, but debian 10 has the stalling issue). This is resolved in this PR by disallowing the use of `ControlMaster` for k8s instances.

<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Manual tests running `sky launch`, `sky exec`, `ssh <cluster-name>`
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

romilbhardwaj commented:
Thanks @landscapepainter! I merged both the latest master and `k8s_cloud_gpu` into `k8s_cloud_beta1`. Can you update this PR by merging the latest `k8s_cloud_beta1` into your branch?

landscapepainter commented:
@romilbhardwaj merged the updated `k8s_cloud_beta1` and added more comments.

One thing that is not resolved is on how to have the users install `socat`. This is not an application that is installed by default and not supported by PYPI. I was wondering if we should inform the users to install it in the document.

Michaelvll commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -4102,11 +4130,16 @@ def _execute_file_mounts(self, handle: CloudVmRayResourceHandle,
-            ip_list, port_list=port_list, **ssh_credentials)
+            ip_list,
+            port_list=port_list,
+            run_on_k8s=run_on_k8s,
```
Instead of explicitly specifying the `run_on_k8s`, it might be better to have this argument generated by `ssh_credentials` argument, where we can check the cloud provider in the YAML file?

Michaelvll replied:
nit for the naming: maybe we can call it `proxy_to_k8s`?

landscapepainter replied:
@Michaelvll Thanks for the direction! This is updated now.

romilbhardwaj commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2824,7 +2824,6 @@ def _sync_workdir(self, handle: CloudVmRayResourceHandle,
                 f'down rsync.{style.RESET_ALL}')
 
         log_path = os.path.join(self.log_dir, 'workdir_sync.log')
-
```
unintended? Same at L4071

romilbhardwaj commented on sky/templates/kubernetes-port-forward-proxy-command.yaml.j2:
```diff
@@ -0,0 +1,25 @@
+set -uo pipefail
+
+# Specifies which config file kubectl should use.
+export KUBECONFIG="{{ kube_config_path }}"
```
Do we need this? We ideally want to operate with whatever `KUBECONFIG` var is set by the user. I know some parts of our code won't support custom `KUBECONFIG`, but we should avoid adding new code that puts strict dependency on a kubeconfig path set by us.

landscapepainter replied:
Gotcha. Got them removed.

romilbhardwaj commented on sky/templates/kubernetes-port-forward-proxy-command.yaml.j2:
```diff
@@ -0,0 +1,25 @@
+    sleep 0.1
+done
+
+# Estalbishes two directional byte streams to handle stdin/stdout between
```
```suggestion
# Establishes two directional byte streams to handle stdin/stdout between
```

romilbhardwaj commented on sky/templates/kubernetes-port-forward-proxy-command.yaml.j2:
```diff
@@ -0,0 +1,25 @@
+
+# Estalbishes two directional byte streams to handle stdin/stdout between
+# terminal and the jump pod
+socat - tcp:{{ ipaddress }}:{{ local_port }}
```
I was trying this our and it silently failed to ssh for a long time before I realized I don't have socat installed.

Is it possible to check if `socat` is installed at the start of the script, raise an error if its not and propagate this error cleanly up to the user? Otherwise, we may want to add a check if `socat` is installed elsewhere in our code...

landscapepainter replied:
I added a check for `socat` installation and displays error msg and exit at the beginning of the script if it's not installed so that it shows the msg when users attempts to `ssh <k8s-instance-name>` without `socat` installed.

But it doesn't seem like there's a clean way to handle this exit and raise an error msg for every possible ssh session runs within skypilot. So I added another check for `socat` installation in `authentication.py/setup_kubernetes_authentication.py` when 'port-forward' mode is being setup.

romilbhardwaj commented on sky/utils/command_runner.py:
```diff
@@ -42,13 +42,16 @@ def _ssh_control_path(ssh_control_filename: Optional[str]) -> Optional[str]:
+def ssh_options_list(
+    ssh_private_key: Optional[str],
+    ssh_control_name: Optional[str],
+    proxy_to_k8s: Optional[bool] = False,
```
Move kwarg `proxy_to_k8s` after `*`

romilbhardwaj replied:
It seems to me the only purpose of this flag `proxy_to_k8s` is to disable ControlMaster. Can we rename this to `disable_control_master` so it's easier to parse and reusable in the future?

romilbhardwaj commented on sky/utils/command_runner.py:
```diff
@@ -79,7 +82,12 @@ def ssh_options_list(ssh_private_key: Optional[str],
     # SSH Control will have a severe delay when using docker_ssh_proxy_command.
     # TODO(tian): Investigate why.
-    if ssh_control_name is not None and docker_ssh_proxy_command is None:
+    # k8s instances are accessed with an ssh session using Proxycommand. The
+    # process running Proxycommand is kept running as long as the ssh session
+    # is running and the ControlMaster keeps the session, which results in
+    # 'ControlPersist' number of seconds delay per ssh commands ran.
```
```suggestion
    # We also do not use ControlMaster when we use `kubectl port-forward` to access  Kubernetes pods over SSH+Proxycommand. This is because the
    # process running ProxyCommand is kept running as long as the ssh session
    # is running and the ControlMaster keeps the session, which results in
    # 'ControlPersist' number of seconds delay per ssh commands ran.
```

romilbhardwaj commented on sky/templates/kubernetes-port-forward-proxy-command.yaml.j2:
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
```
Can you run `pytest tests/test_smoke.py --kubernetes -k "not TestStorageWithCredentials"` to make sure everything, including file_mounts, work correctly? I have manually verified, but going forward we want to run Kubernetes smoke tests for k8s PRs :)

landscapepainter replied:
@romilbhardwaj Currently, passing all the tests besides the ones requiring GPUs for this branch.

romilbhardwaj reviewed (commented):
Thanks @landscapepainter! Works nicely. Left some comments.

I'm planning to refactor the SSH jump pod creation to node_provider in k8s_cloud_beta1 branch, so we may need to update this branch after that's done.

romilbhardwaj reviewed (commented):
@landscapepainter - I have done some refactoring in `k8s_cloud_beta1`. Can you merge those changes?

landscapepainter commented on sky/authentication.py:
```diff
@@ -377,9 +380,57 @@ def setup_scp_authentication(config: Dict[str, Any]) -> Dict[str, Any]:
+                                  ssh_setup_mode: str):
+    """ returns Proxycommand to use when establishing ssh connection
+    to the k8s instance through the jump pod.
+
```
@romilbhardwaj I wrote the doc that can perhaps be a bit long. Please take a look!

landscapepainter reviewed (commented):
@romilbhardwaj Ready for another look!

romilbhardwaj reviewed (commented):
Thanks @landscapepainter! Took another pass at code - will test it soon.

landscapepainter commented:
@romilbhardwaj This is ready for another look!

romilbhardwaj commented:
Thanks @landscapepainter. After #2328 has been merged, I updated `k8s_cloud_beta1` and merged it with the latest master. Before I try this out, can you please update this branch to be up to date with `k8s_cloud_beta1`?

landscapepainter commented:
@romilbhardwaj Merged with updated k8s_cloud_beta1.

Also, I added a functionality to switch the ssh jump pod's service if the already existing service does not match with user's networking configuration in `~/.sky/config.yaml` when `sky launch` is ran. It deletes the existing service and recreates a new service based on the config.

Users can switch between different networking mode by solely recreating the service for ssh jump pod, but this is currently only allowed by running `sky launch`. I'm wondering if we should think about a way to allow users to switch around the modes, i.e. adding another CLI option.

romilbhardwaj commented on sky/authentication.py:
```diff
@@ -404,16 +408,49 @@ def setup_kubernetes_authentication(config: Dict[str, Any]) -> Dict[str, Any]:
+        # the service can be a simple ClusterIP service which we access with
+        # `kubectl port-forward`.
+
+        # Checks if 'socat' is installed
+        try:
+            subprocess.run(['socat', '-V'],
+                           stdout=subprocess.DEVNULL,
+                           stderr=subprocess.DEVNULL,
+                           check=True)
+        except FileNotFoundError:
+            with ux_utils.print_exception_no_traceback():
+                raise RuntimeError(
+                    '`socat` is required to setup Kubernetes cloud with '
+                    '`port-forward` default networking mode and it is not '
+                    'installed. For Debian/Ubuntu system, install it with:\n'
+                    '  $ sudo apt install socat') from None
```
Consider refactor into `kubernetes_utils.check_socat_installed()`

romilbhardwaj commented on sky/utils/kubernetes_utils.py:
```diff
@@ -603,30 +611,110 @@ def __str__(self):
+            'ssh_jump_name': ssh_jump_name,
+            'local_port': ssh_jump_port,
+        }
+        port_forward_proxy_cmd_path = os.path.expanduser(
```
Good to move this `expanduser` to `_get_proxy_command`

romilbhardwaj commented on sky/authentication.py:
```diff
@@ -404,16 +408,49 @@ def setup_kubernetes_authentication(config: Dict[str, Any]) -> Dict[str, Any]:
             raise
 
+    ssh_jump_name = clouds.Kubernetes.SKY_SSH_JUMP_NAME
+    if ssh_setup_mode == 'nodeport':
```
Can we use the enum `KubernetesNetworkingMode` everywhere to avoid hardcoding strings 'nodeport' and 'portforward'? We can also ask the user to use the same string in the config file and then read directly.

landscapepainter replied:
@romilbhardwaj I'm wondering if we should create an `enum` for service types as well since there are currently some places(`setup_kubernetes_authentication`,  `setup_sshjump_svc`) using `NodePort` and `ClusterIP` as a hardcoded string. What do you think?

romilbhardwaj replied:
Yes, I think it'll be good to have `KubernetesServiceType` enum

romilbhardwaj commented on sky/utils/kubernetes_utils.py:
```diff
@@ -603,30 +611,110 @@ def __str__(self):
 
-def get_ssh_proxy_command(private_key_path: str, sshjump_name: str,
-                          namespace: str) -> str:
+def _get_proxy_command(network_mode: KubernetesNetworkingMode,
+                       private_key_path: str,
+                       ssh_jump_port: int,
+                       ssh_jump_ip: str,
+                       port_forward_proxy_cmd_path: str = '') -> str:
+    if network_mode == KubernetesNetworkingMode.NODEPORT:
+        ssh_jump_proxy_command = (f'ssh -tt -i {private_key_path} '
+                                  '-o StrictHostKeyChecking=no '
+                                  '-o UserKnownHostsFile=/dev/null '
+                                  f'-o IdentitiesOnly=yes -p {ssh_jump_port} '
+                                  f'-W %h:%p sky@{ssh_jump_ip}')
+    else:  # network_mode == KubernetesNetworkingMode.PORT_FORWARD:
+        ssh_jump_proxy_command = (
+            f'ssh -tt -i {private_key_path} '
+            f'-o ProxyCommand=\'{port_forward_proxy_cmd_path}\' '
+            '-o StrictHostKeyChecking=no '
+            '-o UserKnownHostsFile=/dev/null '
+            f'-o IdentitiesOnly=yes -p {ssh_jump_port} '
+            f'-W %h:%p sky@{ssh_jump_ip}')
+    return ssh_jump_proxy_command
```
Can this method can be written more generally (`network_mode` and `port_forward_proxy_cmd_path` are redundant)? 

Also consider renaming to `construct_ssh_jump_command` to avoid conflict between name (`_get_proxy_command `) and arg (`port_forward_proxy_cmd_path`)
```suggestion
def construct_ssh_jump_command(private_key_path: str,
                               ssh_jump_port: int,
                               ssh_jump_ip: str,
                               proxy_cmd_path: Optional[str] = None) -> str:
    ssh_jump_proxy_command = (f'ssh -tt -i {private_key_path} '
                              '-o StrictHostKeyChecking=no '
                              '-o UserKnownHostsFile=/dev/null '
                              f'-o IdentitiesOnly=yes -p {ssh_jump_port} '
                              f'-W %h:%p sky@{ssh_jump_ip}')
    if proxy_cmd_path is not None:
        ssh_jump_proxy_command += f'-o ProxyCommand=\'{proxy_cmd_path}\' '
    return ssh_jump_proxy_command
```

romilbhardwaj commented on sky/utils/kubernetes_utils.py:
```diff
@@ -603,30 +611,110 @@ def __str__(self):
+    # ClusterIP service to disallow any ports opened
+    else:
+        ssh_jump_port = LOCAL_PORT_FOR_PORT_FORWARD
+        kube_config_path = os.path.expanduser(kube_config_path)
```
Unused variable (and arg)?

romilbhardwaj commented on tests/test_config.py:
```diff
@@ -9,6 +9,8 @@
 
 VPC_NAME = 'vpc-12345678'
 PROXY_COMMAND = 'ssh -W %h:%p -i ~/.ssh/id_rsa -o StrictHostKeyChecking=no'
+NODEPORT_MODE_NAME = 'nodeport'
+PORT_FORWARD_MODE_NAME = 'port-forward'
```
Consider using enum here

romilbhardwaj reviewed (commented):
Thanks @landscapepainter. Left some comments, still reading code and yet to try.

landscapepainter commented:
Thanks @romilbhardwaj! This is ready for another look.

romilbhardwaj commented on sky/templates/kubernetes-port-forward-proxy-command.sh.j2:
```diff
@@ -0,0 +1,30 @@
+# Establishes two directional byte streams to handle stdin/stdout between
+# terminal and the jump pod.
+# socat process terminates when port-forward terminates.
+socat - tcp:127.0.0.1:{{ local_port }}
```
On running any command (sky launch or ssh), stdout prints:
```
Connection to 127.0.0.1 port 23100 [tcp/*] succeeded!
```
Is there some way to suppress this message?

landscapepainter replied:
Doesn't seem like there is a clear way to go about this especially since we are utilizing socat's stdout/stderr to interact with the node. Wasn't able to find a clear way from [this man page](https://linux.die.net/man/1/socat) as well. Also, it looks like the message is OS specific? I'm using Debian GNU/Linux 10 (buster) and am not seeing the message:

```
$ ssh sky-1aad-gcpuser
Warning: Permanently added '[127.0.0.1]:23100' (ECDSA) to the list of known hosts.
Warning: Permanently added 'xx.xxx.x.x' (ECDSA) to the list of known hosts.
Linux sky-1aad-gcpuser-ray-head 4.19.0-22-cloud-amd64 #1 SMP Debian 4.19.260-1 (2022-09-29) x86_64

The programs included with the Debian GNU/Linux system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Debian GNU/Linux comes with ABSOLUTELY NO WARRANTY, to the extent
permitted by applicable law.
Last login: Tue Aug 29 04:19:04 2023 from xx.xxx.x.x
```

romilbhardwaj commented on sky/templates/kubernetes-port-forward-proxy-command.sh.j2:
```diff
@@ -0,0 +1,30 @@
+fi
+
+# If the port-forward command is already running on the jump pod, kill it
+pgrep -f "kubectl port-forward svc/{{ ssh_jump_name }} {{ local_port }}:22" > /dev/null
+if [ $? -eq 0 ]; then
+    pkill -f "kubectl port-forward svc/{{ ssh_jump_name }} {{ local_port }}:22"
+fi
```
I ran into a bug where creating multiple ssh connections in parallel (e.g., multiple ssh windows or parallel sky launches) would allow only one connection to live at a time. I.e., the previous connection gets abruptly terminated.

To replicate:
1. `sky launch -c test`
2. Open a terminal and run `ssh test`
3. While the previous terminal is running, open a new terminal and run `ssh test`
4. The ssh connection created in step 2 gets broken.

```
(base) sky@test-ray-head:~$ /Users/romilb/.sky/port-forward-proxy-cmd.sh: line 30: 32784 Terminated: 15          kubectl port-forward svc/sky-sshjump-2ea485ef 23100:22
                                           /Users/romilb/.sky/port-forward-proxy-cmd.sh: line 31: kill: (32784) - No such process
     client_loop: send disconnect: Broken pipe
client_loop: send disconnect: Broken pipe
``` 

We may need to 
a) find a better way to handle open `kubectl port-forward` instead of killing them. Perhaps check if it is in use or not
b) If the port is in use, use another local_port to create the new connection

romilbhardwaj replied:
@landscapepainter - here's a script that works and handles multiple SSH connections:

```
#!/usr/bin/env bash
set -uo pipefail

# Checks if socat is installed
if ! command -v socat > /dev/null; then
  echo "Using 'port-forward' mode to run ssh session on Kubernetes instances requires 'socat' to be installed. Please install 'socat'" >&2
  exit
fi

# Checks if lsof is installed
if ! command -v lsof > /dev/null; then
  echo "Checking port availability requires 'lsof' to be installed. Please install 'lsof'" >&2
  exit 1
fi

# Function to check if port is in use
is_port_in_use() {
    local port="$1"
    lsof -i :${port} > /dev/null 2>&1
}

# Start from a fixed local port and increment if in use
local_port={{ local_port }}
while is_port_in_use "${local_port}"; do
    local_port=$((local_port + 1))
done

# Establishes connection between local port and the ssh jump pod
kubectl port-forward svc/{{ ssh_jump_name }} "${local_port}":22 &

# Terminate the port-forward process when this script exits.
K8S_PORT_FWD_PID=$!
trap "kill $K8S_PORT_FWD_PID" EXIT

# checks if a connection to local_port of 127.0.0.1:[local_port] is established
while ! nc -z 127.0.0.1 "${local_port}"; do
    sleep 0.1
done

# Establishes two directional byte streams to handle stdin/stdout between
# terminal and the jump pod.
# socat process terminates when port-forward terminates.
socat - tcp:127.0.0.1:"${local_port}"
```

landscapepainter replied:
Updated the script to the version above and added a check for `lsof` installation in `authentication.py`. Reason why we need to check for installations in both `authentication.py` and `port-fordward-proxy-cmd.sh`: https://github.com/skypilot-org/skypilot/pull/2412#discussion_r1299460245

landscapepainter replied:
@romilbhardwaj Do you think it's necessary to set an upper bound on possible `local_port` usage? It can limit the number of concurrent ssh sessions, but was wondering if we should set a range on available `local_port` for this purpose.

landscapepainter commented:
@romilbhardwaj Ready for another look! Updated the proxy script with what you left in the comment and creaed another `enum` for service types. Also added another check for `lsof` installation in `kubernetes_utils.py`

romilbhardwaj commented on sky/utils/command_runner.py:
```diff
@@ -198,7 +213,8 @@ def make_runner_list(
             port_list = [22] * len(ip_list)
         return [
             SSHCommandRunner(ip, ssh_user, ssh_private_key, ssh_control_name,
```
On GKE clusters, This PR is currently using the external IP of nodes (`sky@34.16.78.81`):
```
2023-08-29 17:53:01,300	VVINFO command_runner.py:367 -- Full command is `[1mssh -tt -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_8d2ba49c09/098f6bcd46/%C -o ControlPersist=10s -o ProxyCommand=ssh -tt -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -p 23100 -W %h:%p sky@34.16.78.81 -o ProxyCommand='/Users/romilb/.sky/port-forward-proxy-cmd.sh'  -o Port=22 -o ConnectTimeout=120s sky@10.72.0.17 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (mkdir -p ~/.sky/.runtime_files)'[22m[26m`
building file list ... done
```
This will not work if the nodes are behind a firewall. It should instead of be connecting to `sky@127.0.0.1` since the the port-forward is running locally. 

You may need to update get_external_ip to accept an arg with the `KubernetesNetworkingMode`, which would return 127.0.0.1 if the mode is port-forward, else use the existing logic if the mode is nodeport.

landscapepainter reviewed (commented):
@romilbhardwaj This is ready for another look.
- merged [@hemildesai's branch](https://github.com/landscapepainter/skypilot/pull/1) on allowing k8s specific env. vars accessible within ssh session and updated the implementation
- updated `get_external_ip`
- removed `query_env_vars` and `_update_envs_for_k8s`

Confirmed successful provision on k8s instance with GPU

romilbhardwaj commented on sky/skylet/providers/kubernetes/node_provider.py:
```diff
@@ -237,10 +239,42 @@ def create_node(self, node_config, tags, count):
+        # shell sessions.
+        set_k8s_env_var_cmd = [
+            '/bin/sh', '-c',
+            'printenv | awk \'{print "export " $0}\' > ~/k8s_env_var.sh && sudo mv ~/k8s_env_var.sh /etc/profile.d/k8s_env_var.sh'
```
Bug: After the first run, `bash: 419: No such file or directory` gets printed every time I SSH or exec a task.

Repro:
1. `sky launch -c test --gpus T4:1 -- nvidia-smi`
2. `sky exec test --gpus T4:1 -- nvidia-smi` (This prints the above error). [Logs](https://gist.github.com/romilbhardwaj/76c5e0cbfc59623d4e9f3b1f5776ade7).
3. `ssh test` this also prints the above error.

Cause:
`/etc/profile.d/k8s_env_var.sh` contains a line like:
```
export NVIDIA_REQUIRE_CUDA=cuda>=11.6 brand=tesla,driver>=418,driver<419 brand=tesla,driver>=450,driver<451 brand=tesla,driver>=470,driver<471 brand=unknown,driver>=470,driver<471 brand=nvidia,driver>=470,driver<471 brand=nvidiartx,driver>=470,driver<471 brand=geforce,driver>=470,driver<471 brand=geforcertx,driver>=470,driver<471 brand=quadro,driver>=470,driver<471 brand=quadrortx,driver>=470,driver<471 brand=titan,driver>=470,driver<471 brand=titanrtx,driver>=470,driver<471
```
The shell tries to interpret special characters > and <, causing errors.

Solution - wrap the exported value in single quotes:
```
export NVIDIA_REQUIRE_CUDA='cuda>=11.6 brand=tesla,driver>=418,driver<419 brand=tesla,driver>=450,driver<451 brand=tesla,driver>=470,driver<471 brand=unknown,driver>=470,driver<471 brand=nvidia,driver>=470,driver<471 brand=nvidiartx,driver>=470,driver<471 brand=geforce,driver>=470,driver<471 brand=geforcertx,driver>=470,driver<471 brand=quadro,driver>=470,driver<471 brand=quadrortx,driver>=470,driver<471 brand=titan,driver>=470,driver<471 brand=titanrtx,driver>=470,driver<471'
```

Catch - if the envvar var value itself contains single quotes, we'll need to be careful in handling single quotes (`'`) appearing in the string. See some suggestions [here](https://stackoverflow.com/a/1315213).

romilbhardwaj replied:
`sudo` may not be present on all images. Can we try mv without sudo first, and use sudo if it fails?

romilbhardwaj commented:
Thanks @landscapepainter - can you also run `pytest tests/test_smoke.py --kubernetes -k "not TestStorageWithCredentials"` on a [test GKE cluster with all resources required for smoke tests](https://github.com/skypilot-org/skypilot/blob/master/tests/kubernetes/README.md#running-a-gke-cluster)?
